File: python/app/filters/noise_filters.py
"""Noise generation and reduction filters."""

import cv2
import numpy as np
from typing import List

from app.filters.base_filter import ImageFilter
from app.pfm import PFMSetting, SettingType


class GaussianNoiseFilter(ImageFilter):
    @property
    def name(self) -> str: return "Add Gaussian Noise"
    @property
    def category(self) -> str: return "Noise"

    def _define_settings(self) -> List[PFMSetting]:
        return [
            PFMSetting("mean", "Mean", SettingType.NUMBER, 0.0, -100.0, 100.0, 1.0),
            PFMSetting("sigma", "Std Dev", SettingType.NUMBER, 25.0, 0.0, 100.0, 1.0)
        ]

    def process(self, image: np.ndarray) -> np.ndarray:
        mean = self.get("mean")
        sigma = self.get("sigma")
        gauss = np.random.normal(mean, sigma, image.shape)
        noisy = image.astype(np.float32) + gauss
        return np.clip(noisy, 0, 255).astype(np.uint8)


class SaltAndPepperFilter(ImageFilter):
    @property
    def name(self) -> str: return "Salt & Pepper Noise"
    @property
    def category(self) -> str: return "Noise"

    def _define_settings(self) -> List[PFMSetting]:
        return [
            PFMSetting("prob", "Probability", SettingType.PERCENTAGE, 5.0, 0.0, 100.0, 1.0)
        ]

    def process(self, image: np.ndarray) -> np.ndarray:
        prob = self.get("prob") / 100.0
        noisy = np.copy(image)
        # Salt
        num_salt = np.ceil(prob * image.size * 0.5)
        coords = [np.random.randint(0, i - 1, int(num_salt)) for i in image.shape]
        noisy[tuple(coords)] = 255
        # Pepper
        num_pepper = np.ceil(prob * image.size * 0.5)
        coords = [np.random.randint(0, i - 1, int(num_pepper)) for i in image.shape]
        noisy[tuple(coords)] = 0
        return noisy


class DenoiseFilter(ImageFilter):
    @property
    def name(self) -> str: return "Denoise (Non-Local Means)"
    @property
    def category(self) -> str: return "Noise"

    def _define_settings(self) -> List[PFMSetting]:
        return [
            PFMSetting("h", "Strength (h)", SettingType.NUMBER, 10.0, 1.0, 50.0, 1.0)
        ]

    def process(self, image: np.ndarray) -> np.ndarray:
        h = self.get("h")
        if len(image.shape) == 3:
            return cv2.fastNlMeansDenoisingColored(image, None, h, h, 7, 21)
        else:
            return cv2.fastNlMeansDenoising(image, None, h, 7, 21)

File: cpp/filters/noise_filters.cpp
#include "filters/noise_filters.h"
#include <opencv2/imgproc.hpp>
#include <opencv2/photo.hpp>

// ---------------------------------------------------------------------------
// GaussianNoiseFilter
// ---------------------------------------------------------------------------
QVector<PFMSetting> GaussianNoiseFilter::defineSettings() const {
    return {
        { "mean", "Mean", SettingType::Number, 0.0, QVariant(), -100.0, 100.0, -100.0, 100.0, 1.0 },
        { "sigma", "Std Dev", SettingType::Number, 25.0, QVariant(), 0.0, 100.0, 0.0, 100.0, 1.0 }
    };
}

cv::Mat GaussianNoiseFilter::process(const cv::Mat& image) {
    double mean = get("mean").toDouble();
    double sigma = get("sigma").toDouble();

    cv::Mat noise(image.size(), image.type());
    cv::randn(noise, mean, sigma);
    
    cv::Mat img16;
    image.convertTo(img16, CV_16S);
    
    cv::Mat noise16;
    noise.convertTo(noise16, CV_16S);
    
    img16 += noise16;
    
    cv::Mat out;
    img16.convertTo(out, CV_8U);
    return out;
}

// ---------------------------------------------------------------------------
// SaltAndPepperFilter
// ---------------------------------------------------------------------------
QVector<PFMSetting> SaltAndPepperFilter::defineSettings() const {
    return {
        { "prob", "Probability", SettingType::Percentage, 5.0, QVariant(), 0.0, 100.0, 0.0, 100.0, 1.0 }
    };
}

cv::Mat SaltAndPepperFilter::process(const cv::Mat& image) {
    double prob = get("prob").toDouble() / 100.0;
    if (prob <= 0.0) return image.clone();

    cv::Mat out = image.clone();
    cv::Mat noise(image.size(), CV_32F);
    cv::randu(noise, 0.0, 1.0);

    // Apply pepper
    out.setTo(cv::Scalar::all(0), noise < (prob / 2.0));
    // Apply salt
    out.setTo(cv::Scalar::all(255), (noise >= (prob / 2.0)) & (noise < prob));
    
    return out;
}

// ---------------------------------------------------------------------------
// DenoiseFilter
// ---------------------------------------------------------------------------
QVector<PFMSetting> DenoiseFilter::defineSettings() const {
    return {
        { "h", "Strength (h)", SettingType::Number, 10.0, QVariant(), 1.0, 50.0, 1.0, 50.0, 1.0 }
    };
}

cv::Mat DenoiseFilter::process(const cv::Mat& image) {
    double h = get("h").toDouble();
    cv::Mat out;
    if (image.channels() == 3) {
        cv::fastNlMeansDenoisingColored(image, out, h, h, 7, 21);
    } else if (image.channels() == 1) {
        cv::fastNlMeansDenoising(image, out, h, 7, 21);
    } else {
        out = image.clone();
    }
    return out;
}

// ---------------------------------------------------------------------------
// GaussianNoise2Filter
// ---------------------------------------------------------------------------
cv::Mat GaussianNoise2Filter::process(const cv::Mat& image) {
    cv::Mat noise(image.size(), image.type());
    cv::randn(noise, 0.0, 10.0);
    
    cv::Mat img16;
    image.convertTo(img16, CV_16S);
    
    cv::Mat noise16;
    noise.convertTo(noise16, CV_16S);
    
    img16 += noise16;
    
    cv::Mat out;
    img16.convertTo(out, CV_8U);
    return out;
}

// ---------------------------------------------------------------------------
// SpeckleNoiseFilter
// ---------------------------------------------------------------------------
cv::Mat SpeckleNoiseFilter::process(const cv::Mat& image) {
    cv::Mat noise(image.size(), CV_32F);
    cv::randn(noise, 0.0, 1.0);
    
    cv::Mat img32;
    image.convertTo(img32, CV_32F);
    
    cv::Mat out = img32 + img32.mul(noise) * 0.1;
    out.convertTo(out, CV_8U);
    return out;
}
